Sign up feature (frontend only)

EvanAntoni opened the pull request:
Form with input and submit handlers

EvanAntoni committed:
add sign up form

EvanAntoni committed:
add input and form submit handlers

EvanAntoni committed:
add post-submit components

EvanAntoni committed:
add typeface-roboto font

EvanAntoni committed:
edit text in Sign Up button

serge30 commented on frontend/src/components/SignUp/SignUpForm.jsx:
```diff
@@ -0,0 +1,159 @@
+  handleSubmit(e) {
+    e.preventDefault();
+    const {name, email, password} = this.state;
+    const formData = {name, email, password};
```
Client side validation is needed here. With clear messages what is wrong.

EvanAntoni replied:
Done. Review, please.

EvanAntoni committed:
add form validation

VitaliiBondar reviewed (commented):
You should install "react-material-ui-form-validator" before use

VitaliiBondar commented:
![tomuthhome](https://user-images.githubusercontent.com/45948982/51200814-3669a400-1903-11e9-86e8-64dcb1d66c0b.png)

I think that there are too many links to the Home page, maybe it's better to make a button "back"

VitaliiBondar commented on frontend/src/components/SignUp/SignUpForm.jsx:
```diff
@@ -0,0 +1,192 @@
+          </Typography>
+          <Grid container direction='column' alignContent='center'>
+            <Grid item>
+              <Button variant="outlined" href="/" className={classes.button}>
```
Do not use ..Button href="/".. to jump to another page. Our project is a SPA. Use the Link component from "react-router-dom".

EvanAntoni replied:
Why Button can't be used in this case? What are the benefits of using Link instead of href?

VitaliiBondar replied:
When you use ..Button href="/".. , it reload a page... Try something like "..Button component={Link} to={'/'} "

EvanAntoni replied:
I got it. Good advice. Thanks.

VitaliiBondar commented on frontend/src/components/SignUp/SignUpForm.jsx:
```diff
@@ -0,0 +1,192 @@
+                      label="Password"
+                      onChange={this.handleChange}
+                      className={classes.textField}
+                      validators={['required', "minStringLength:8", "maxStringLength:16", "matchRegexp:^.*[a-zA-Z]{2}"]}
```
If you use it, you need to write a description for the password field. Something like "Password must be minimum 8 characters, including at least one uppercase letter, one lowercase and one number."

VitaliiBondar replied:
![pass](https://user-images.githubusercontent.com/45948982/51209311-a171a580-1918-11e9-9a6d-5585e072ab08.png)
like this

serge30 replied:
I don't like the idea to restrict users with max length of password. What if user has as a password little poem? ;) Anyway we will hash it, and the size it occupy in DB will be always the same. I believe you may restrict it to something like 256 symbols.

VitaliiBondar replied:
You use matchRegexp:^.*[a-zA-Z]{2}, but if the user has password like "СуперПароль" ?

EvanAntoni replied:
> If you use it, you need to write a description for the password field. Something like "Password must be minimum 8 characters, including at least one uppercase letter, one lowercase and one number."

Ok. I will add helper text.

> I don't like the idea to restrict users with max length of password. What if user has as a password little poem? ;)

No problem. I will correct it.

> You use matchRegexp:^.*[a-zA-Z]{2}, but if the user has password like "СуперПароль" ?

The user will see the inscription "Please use LATIN symbols and at least 2 letters". But I will consider this and add "а-яА-Я" symbols too. I think there should be no problems.

VitaliiBondar commented on frontend/src/components/SignUp/SignUpForm.jsx:
```diff
@@ -0,0 +1,192 @@
+  },
+});
+
+class SubmittedForm extends React.Component {
```
I think it would be better to make it a separate component

EvanAntoni replied:
OK)

EvanAntoni commented:
> You should install "react-material-ui-form-validator" before use

I forgot to add the file to the commit. Thanks.

> I think that there are too many links to the Home page, maybe it's better to make a button "back"

I planned to fix the header in the future because at this stage I do not completely understand how to do it correctly.
It is very strange that the "GO TO HOMEPAGE" button has shifted to the left. My Firefox Browser renders it in the center.

EvanAntoni committed:
add form validator and roboto font

EvanAntoni committed:
create the SubmittedForm component, fix button on this component, fix password fields properties

Create a separate SubmittedForm component, fix button link and position on this component.
Add helper text for the password field.
Remove password restriction by characters.
Add support for Cyrillic characters.

serge30 commented on frontend/src/components/SignUp/SignUpForm.jsx:
```diff
@@ -0,0 +1,172 @@
+                      label="Password"
+                      onChange={this.handleChange}
+                      className={classes.textField}
+                      validators={['required', "minStringLength:8", "matchRegexp:^.*[a-zA-Zа-яА-Я]{2}"]}
```
Please try to avoid using non-ascii symbols in code. There is recommendation from MDN how to use unicode symbols in  regexps: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp#Regular_expression_and_Unicode_characters

EvanAntoni replied:
Thanks for the information. I want to make some changes:

1. I want to use this RegExp `.*[a-zA-Z\u0430-\u044f\u0410-\u042f].*` (any characters in the password, but must contain at least one letter).
2. I want to edit restriction for name (2 characters instead of 3)
3. I want to show error after unsuccessful form submission on the same page.
4. I want to replace ordinary functions with arrows functions.

What do you think about it all?

serge30 replied:
No objections from my side

viktormovchaniuk replied:
Does we need restrict min length on name? yesterday, on demo, we decide just make it field required as i understood

serge30 replied:
@viktormovchaniuk yep, you are right

EvanAntoni replied:
> Does we need restrict min length on name? yesterday, on demo, we decide just make it field required as i understood

Because of our Chinese users, where one symbol can mean one sentence. But I thought we don't orient at them.

EvanAntoni replied:
I already removed min length restriction.

Be cool, be nice
Removed restrictions, guys

EvanAntoni committed:
edit password validation, remove name restriction and roboto font

Edit password validation (regexp) and password helper text.
Remove name restriction by characters.
Remove Roboto fonts from dependencies in package.json

EvanAntoni committed:
edit unsuccessful form submit message

EvanAntoni committed:
replace usual functions to arrow functions

EvanAntoni committed:
Merge branch 'develop' into sign_up_feature
